File: Javascript/NQueens.js
// Place N queens on an N×N chessboard so no two queens attack each other. Return one possible solution.

function NQueens(n) {
    const cols = new Set();
    const diag1 = new Set(); 
    const diag2 = new Set(); 
    let res = 0;

    function backtrack(row) {
        if (row === n) {
            res++;
            return;
        }

        for (let col = 0; col < n; col++) {
            if (cols.has(col)) continue;
            if (diag1.has(row - col)) continue;
            if (diag2.has(row + col)) continue;

            // Place queen
            cols.add(col);
            diag1.add(row - col);
            diag2.add(row + col);

            backtrack(row + 1);

            // Remove queen & recurse
            cols.delete(col);
            diag1.delete(row - col);
            diag2.delete(row + col);
        }
    }

    backtrack(0);
    return res;
}

const n = 8;
console.log(`Number of solutions for ${n}-Queens:`, NQueens(n));

File: C++/NQueens.cpp
// Place N queens on an N×N chessboard so no two queens attack each other. Return one possible solution.
#include <iostream>
#include <unordered_set>
#include <vector>
using namespace std;

class Solution{
public:
    std::unordered_set<int> cols;
    std::unordered_set<int> diag1;
    std::unordered_set<int> diag2;
    int res;
    int n;

    int NQueens(int n){
        this->n = n;
        res = 0;

        // Reset
        cols.clear();
        diag1.clear();
        diag2.clear();
        
        Recurs(0);
        return res;
    }

    void Recurs(int rows){
        if(rows == n){
            res++;
            return;
        }

        for(int i = 0; i < n; i++){
            if(cols.find(i) != cols.end()) continue;
            if(diag1.find(rows - i) != diag1.end()) continue;
            if(diag2.find(rows + i) != diag2.end()) continue;

            // Place queen
            cols.insert(i);
            diag1.insert(rows - i);
            diag2.insert(rows + i);

            // Recurse & remove queen
            Recurs(rows++);
            cols.erase(i);
            diag1.erase(rows - i);
            diag2.erase(rows + i);
        }
    }
};

int main(){
    Solution sol;
    int n = 8;
    std::cout <<"Number of solutions for " << n << " -Queens: " << sol.NQueens(n) << std::endl;
}